File: 251004/Kalitsko/3lab/discussion/src/main/java/com/example/rest/controller/PostController.java
package com.example.rest.controller;

import com.example.rest.dto.PostRequestTo;
import com.example.rest.dto.PostResponseTo;
import com.example.rest.dto.PostUpdate;
import com.example.rest.service.Impl.PostService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;


@RestController
@RequestMapping("/api/v1.0/posts")
public class PostController {

    private final PostService postService;

    @Autowired
    public PostController(PostService postService) {
        this.postService = postService;
    }

    // 1. Создание нового поста
    @PostMapping
    public ResponseEntity<PostResponseTo> createPost(@RequestBody PostRequestTo requestTo) {
        PostResponseTo responseTo = postService.createPost(requestTo);
        return ResponseEntity.ok(responseTo);
    }

    // 2. Получение всех постов
    @GetMapping
    public ResponseEntity<List<PostResponseTo>> getAllPosts() {
        List<PostResponseTo> posts = postService.getAllPosts();
        return ResponseEntity.ok(posts);
    }

    // 3. Получение поста по ключу
    @GetMapping("{id}")
    public ResponseEntity<PostResponseTo> getPostById(@PathVariable Long id) {

        PostResponseTo responseTo = postService.getPostById(id);
        return ResponseEntity.ok(responseTo);
    }

    // 4. Обновление поста
    @PutMapping
    public ResponseEntity<PostResponseTo> updatePost(@Valid @RequestBody PostUpdate postUpdate) {

        PostResponseTo responseTo = postService.updatePost(postUpdate);
        return ResponseEntity.ok(responseTo);
    }

    // 5. Удаление поста
    @DeleteMapping("{id}")
    public ResponseEntity<Void> deletePost(@PathVariable Long id) {

        postService.deletePost(id);
        return ResponseEntity.noContent().build();
    }
}

File: 251003/Pankratiev/rest/publisher/src/main/kotlin/bsuir/dc/publisher/controller/PostController.kt
package bsuir.dc.publisher.controller

import bsuir.dc.publisher.dto.from.PostRequestTo
import bsuir.dc.publisher.dto.to.PostResponseTo
import bsuir.dc.publisher.service.PostService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.reactive.function.client.WebClient

@RestController
@RequestMapping("/api/v1.0/posts")
class PostController(
    private val postService: PostService,
    private val webClient: WebClient
) {

    @PostMapping
    fun createPost(@RequestBody @Valid postRequestTo: PostRequestTo): ResponseEntity<PostResponseTo> {
        val createdPost = postService.createPost(postRequestTo)
        sendPostRequest(createdPost)
        return ResponseEntity(createdPost, HttpStatus.CREATED)
    }

    @GetMapping("/{id}")
    fun getPostById(@PathVariable id: Long): ResponseEntity<PostResponseTo> {
        val post = postService.getPostById(id)
        return ResponseEntity.ok(post)
    }

    @GetMapping
    fun getAllPosts(): ResponseEntity<List<PostResponseTo>> {
        val posts = postService.getAllPosts()
        return ResponseEntity.ok(posts)
    }

    @PutMapping()
    fun updatePost(@RequestBody @Valid postRequestTo: PostRequestTo): ResponseEntity<PostResponseTo> {
        val updatedPost = postService.updatePost(postRequestTo.id, postRequestTo)
        sendPutRequest(updatedPost)
        return ResponseEntity.ok(updatedPost)
    }

    @DeleteMapping("/{id}")
    fun deletePost(@PathVariable id: Long): ResponseEntity<Void> {
        sendDeleteRequest(id)
        postService.deletePost(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/by-issue/{issueId}")
    fun getPostsByIssueId(@PathVariable issueId: Long): ResponseEntity<List<PostResponseTo>> {
        val posts = postService.getPostsByIssueId(issueId)
        return ResponseEntity.ok(posts)
    }

    private val url = "http://localhost:24130/api/v1.0/posts"

    fun sendPostRequest(body: PostResponseTo) {
        webClient.post()
            .uri(url)
            .bodyValue(body)
            .retrieve()
            .toBodilessEntity()
            .subscribe()
    }

    fun sendPutRequest(body: PostResponseTo) {
        webClient.put()
            .uri(url)
            .bodyValue(body)
            .retrieve()
            .toBodilessEntity()
            .subscribe()
    }

    fun sendDeleteRequest(id: Long) {
        webClient.delete()
            .uri("$url/$id")
            .retrieve()
            .toBodilessEntity()
            .subscribe()
    }
}